libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m

```diff
@@ -235,27 +250,37 @@ - (void)replayRequestIfRequired:(NSData *)data response:(NSURLResponse *)respons
     NSInteger statusCode = [(NSHTTPURLResponse *)response statusCode];
     if (statusCode == 401 || statusCode == 403) {
         if (shouldRetry) {
-            SFUserAccount *user = [SFUserAccountManager sharedInstance].currentUser;
             [SFSDKCoreLogger i:[self class] format:@"%@: REST request failed due to expired credentials. Attempting to refresh credentials.", NSStringFromSelector(_cmd)];
-            self.oauthSessionRefresher = [[SFOAuthSessionRefresher alloc] initWithCredentials:user.credentials];
-            __weak __typeof(self) weakSelf = self;
-            [self.oauthSessionRefresher refreshSessionWithCompletion:^(SFOAuthCredentials *updatedCredentials) {
-                __strong typeof(weakSelf) strongSelf = weakSelf;
-                [SFSDKCoreLogger i:[strongSelf class] format:@"%@: Credentials refresh successful. Replaying original REST request.", NSStringFromSelector(_cmd)];
-                [strongSelf send:request delegate:delegate shouldRetry:NO];
-            } error:^(NSError *refreshError) {
-                __strong typeof(weakSelf) strongSelf = weakSelf;
-                [SFSDKCoreLogger e:[strongSelf class] format:@"Failed to refresh expired session. Error: %@", refreshError];
-                if ([refreshError.domain isEqualToString:kSFOAuthErrorDomain] && refreshError.code == kSFOAuthErrorInvalidGrant) {
-                    [SFSDKCoreLogger i:[strongSelf class] format:@"%@ Invalid grant error received, triggering logout.", NSStringFromSelector(_cmd)];
-
-                    // Make sure we call logout on the main thread.
-                    dispatch_async(dispatch_get_main_queue(), ^{
-                        [strongSelf createAndStoreLogoutEvent:error user:user];
-                        [[SFAuthenticationManager sharedManager] logoutUser:user];
-                    });
-                }
-            }];
+            SFUserAccount *user = [SFUserAccountManager sharedInstance].currentUser;
+
+            /*
+             * Sends the session refresh request if an OAuth session is not being refreshed.
+             * Otherwise, wait for the current session refresh call to complete before sending.
+             */
+            if (!self.sessionRefreshInProgress) {
```

Comment:
It seems like you still want to synchronize or serial dispatch this block of code.  Otherwise, you have a race condition with multiple requests potentially entering what should otherwise be an exclusive block.